src/databricks/labs/ucx/source_code/linters/graph_walkers.py

```diff
@@ -36,22 +36,37 @@
 
 
 class DependencyGraphWalker(abc.ABC, Generic[T]):
+    """Walks over the dependencies in a graph starting from the root dependencies going depth first.
 
-    def __init__(self, graph: DependencyGraph, walked_paths: set[Path], path_lookup: PathLookup):
+    Implemented as an object to iterate over, for example:
+    ``` python
+    walker = DependencyGraphWalker()
+    for processed_dependency_output in walker:
+        # Do something with output
+    ```
+    """
+
+    def __init__(self, graph: DependencyGraph, path_lookup: PathLookup):
         self._graph = graph
-        self._walked_paths = walked_paths
         self._path_lookup = path_lookup
-        self._lineage: list[Dependency] = []
+
+        self._walked_paths = set[Path]()
+        self._lineage = list[Dependency]()
 
     def __iter__(self) -> Iterator[T]:
+        """Iterate over the dependencies starting from the root."""
         for dependency in self._graph.root_dependencies:
             # the dependency is a root, so its path is the one to use
             # for computing lineage and building python global context
-            root_path = dependency.path
-            yield from self._iter_one(dependency, self._graph, root_path)
+            yield from self._iter_one(dependency, self._graph, dependency.path)
 
     def _iter_one(self, dependency: Dependency, graph: DependencyGraph, root_path: Path) -> Iterable[T]:
+        """Iterate over a single dependency going depth first."""
         if dependency.path in self._walked_paths:
+            # TODO:
```

Comment:
This is an open question to me, @pritishpai and @asnare , please have a look.

I am inclined to leave the question open but not remove the logic to be on the safer side until we answered the open question. 

Note that removing this logic does **NOT** make any unit test fail